sql/cagg_api.sql

```diff
@@ -0,0 +1,136 @@
+      WITH
+         -- Collect ranges from the invalidation log and convert them
+         -- to correct type.
+	 timestamps AS MATERIALIZED (
+	    SELECT to_timestamp(lowest_modified_value) AS start,
+		   to_timestamp(greatest_modified_value) AS finish
+	      FROM continuous_aggs_hypertable_invalidation_log
+	     WHERE hypertable_id = l_hypertable_id
+	 ),
+	 -- Since a range can end at the end of a bucket and this
+	 -- range should not include the next bucket, we need to
+	 -- subtract one microsecond before computing the bucket and
+	 -- then add the width again.
+	 ranges AS MATERIALIZED (
+	   SELECT width,
+		  tstzrange(time_bucket(width, start),
+		  	    time_bucket(width, finish - '1 microsecond'::interval) + width) AS bucket
+	     FROM timestamps CROSS JOIN UNNEST(bucket_widths) w(width)
+	 )
+      SELECT width, pg_current_snapshot()::text, range_agg(bucket) ranges
+        FROM ranges GROUP BY width
+   );
+END
+$body$
+LANGUAGE plpgsql
+SET search_path = _timescaledb_catalog, _timescaledb_functions, public;
+
+CREATE OR REPLACE FUNCTION _timescaledb_functions.get_hypertable_invalidations(
+   hypertable REGCLASS,
+   base TIMESTAMP,
+   bucket_widths INTERVAL[]
+) RETURNS TABLE (bucket_width INTERVAL, token TEXT, invalidations TSMULTIRANGE) AS
+$body$
+DECLARE
+   l_hypertable_id INTEGER := get_hypertable_info(hypertable, 'timestamp'::regtype);
+BEGIN
+   RETURN QUERY (
+      WITH
+         -- Collect ranges from the invalidation log and convert them
+         -- to correct type.
+	 timestamps AS MATERIALIZED (
+	    SELECT to_timestamp_without_timezone(lowest_modified_value) AS start,
+		   to_timestamp_without_timezone(greatest_modified_value) AS finish
+	      FROM continuous_aggs_hypertable_invalidation_log
+	     WHERE hypertable_id = l_hypertable_id
+	 ),
+	 -- Compute bucket-aligned ranges from the ranges above. Since
+	 -- a range can end at the end of a bucket and this range
+	 -- should not include the next bucket, we need to subtract
+	 -- one microsecond before computing the bucket and then add
+	 -- the width again.
+	 ranges AS MATERIALIZED (
+	   SELECT width,
+		  tsrange(time_bucket(width, start),
+		          time_bucket(width, finish - '1 microsecond'::interval) + width) AS bucket
+	     FROM timestamps CROSS JOIN UNNEST(bucket_widths) w(width)
+	 )
+      SELECT width, pg_current_snapshot()::text, range_agg(bucket)
+        FROM ranges GROUP BY width
+   );
+END
+$body$
+LANGUAGE plpgsql
+SET search_path = _timescaledb_catalog, _timescaledb_functions, public;
+
+-- Accept a set of hypertable invalidations.
+--
+-- This procedure is used to accept all invalidations using the token
+-- returned by a previous call of get_hypertable_invalidations().
+CREATE OR REPLACE PROCEDURE _timescaledb_functions.accept_hypertable_invalidations(
+       token TEXT
+) AS
+$body$
+BEGIN
+   DELETE FROM continuous_aggs_hypertable_invalidation_log
+    WHERE pg_visible_in_snapshot(xmin::text::xid8, token::pg_snapshot);
```

Comment:
Shoudn't we store the ranges produced into the materialization invalidation logs and then remove it from hypertable invalidation log?